### PR TITLE
[BFD]Retry create BFD with different source UDP port on failure

### PR DESCRIPTION
**Pull request by dgsudharsan**
**What I did**
The create_bfd_session API can fail when a source port already used by a different application is passed to SAI create. To avoid failure, a retry logic is added to try with a different port number three times before failing

**Why I did it**
To avoid BFD failure and orchagent crash due to port conflict.

**How I verified it**
Verified it manually by having an application binding the port and checking if retry is done when BFD tries the same port

**Details if related**

### CODE REVIEW

File: orchagent/bfdorch.cpp [hunk 1]
```diff
@@ -410,12 +430,24 @@ bool BfdOrch::create_bfd_session(const string& key, const vector<FieldValueTuple
...
+        attr_idx++;
     }
 
     fvVector.emplace_back("state", session_state_lookup.at(SAI_BFD_SESSION_STATE_DOWN));
 
     sai_object_id_t bfd_session_id = SAI_NULL_OBJECT_ID;
-    sai_status_t status = sai_bfd_api->create_bfd_session(&bfd_session_id, gSwitchId, (uint32_t)attrs.size(), attrs.data());
+    sai_status_t status = sai_bfd_api->create_bfd_session(&bfd_session_id, gSwitchId, attr_idx, attrs.data());
+
+    for (int retry = 0; (status != SAI_STATUS_SUCCESS) && (retry < NUM_BFD_SRCPORT_RETRIES); retry++)
```
Comment (prsunny) [inline orchagent/bfdorch.cpp:441]: This is a corner case scenario. I think we should handle it in a separate function and invoke if status is Failure.  You can loop the attributes to find the src port attr index.  Managing index by attr_id++ is also error prone when someone adds a new attribute.

Comment (dgsudharsan) [reply orchagent/bfdorch.cpp]: Updated

File: orchagent/bfdorch.cpp [hunk 2]
```diff
@@ -439,6 +446,38 @@ bool BfdOrch::create_bfd_session(const string& key, const vector<FieldValueTuple
...
+sai_status_t BfdOrch::retry_create_bfd_session(sai_object_id_t &bfd_session_id, vector<sai_attribute_t> attrs)
+{
+    sai_status_t status = SAI_STATUS_FAILURE;
+
+    for (int retry = 0; retry < NUM_BFD_SRCPORT_RETRIES; retry++)
+    {
+        update_port_number(attrs);
+        status = sai_bfd_api->create_bfd_session(&bfd_session_id, gSwitchId,
+                                                 (uint32_t)attrs.size(), attrs.data());
+        if(status == SAI_STATUS_SUCCESS)
```
Comment (prsunny) [inline orchagent/bfdorch.cpp:473]: space after if

**Review by prsunny — COMMENTED** [review]
are you planning to add any VS test. code coverage _may_ fail if the additional code is not covered by test

File: orchagent/bfdorch.cpp (hunk 2 above)
Comment (dgsudharsan) [reply orchagent/bfdorch.cpp]: Updated

**Comment by liat-grozovik** [comment]
LGTM. small comment
@prsunny  adding a code coverage for that will requires mock and not regular functional test. we need to make sure the port is used by other application and then try to create bfd. Do you still think adding a mock in the VS test is the right way? it is not unit test.

**Review by prsunny — APPROVED** [review]

**Comment by prsunny** [comment]
@judyjoseph , can you please cherry-pick to 202111?

**Pull request merged**